### PR TITLE
[RISCV][clang] Don't enable -mrelax-all for -O0 on RISC-V

### LINKED ISSUE #87127: [RISCV] Relocation truncated to fit with -O0 on very large functions.
clang -O0 implicitly passes -mrelax-all. This causes the MC layer to relax all jumps to the longest form possible. Except we don't convert uncompressed relative jumps to indirect since that requires a temporary register.

The CodeGen layer carefully measured the function size and relaxed any branches and unconditional branches that it needed to, but it is not aware of the relaxation the MC layer is going to do for -mrelax-all.

If the function is sufficiently large, relaxing the branches in the MC layer may cause some unconditional jump to go out of range leading to a linking failure.

I believe the branch relaxation was added to the MC layer to make things easier for programmers writing handwritten assembly so they didn't have to worry about the small offset of conditional branches. Maybe we should disable it when assembling CodeGen output?

cc: @preames @MaskRay @jrtc27 @kito-cheng @asb

### PR DESCRIPTION
**Pull request by topperc**
-O0 implies -mrelax-all as an assembler compile time optimization. -mrelax-all allows the assembler to complete layout in 2 passes instead of doing iterative branch relaxation.

Jump offsets larger than +/-1MiB require an indirect jump on RISC-V. This can't be done by the assembler, so we use a branch relaxation MIR pass and use register scavenging to find a free register.

The conditional branch offsets for RISC-V are also somewhat small so we support MC layer branch relaxation to make life easier for assembly programmers. This may also cover up bugs in our function size estimation in MachineIR.

Enabling -mrelax-all causes the MC layer relaxation to agressively relax branches. This increases code size and can create cases where we need an indirect jump, but we can't create one. This leads to linker failures.

The easiest way to avoid this is to not default to -mrelax-all for -O0 and sacrifice the compile time optimization. That's what this patch does.

Fixes #87127

### CODE REVIEW

**Review by MaskRay — APPROVED** [review]
I left a comment at 
https://github.com/llvm/llvm-project/issues/87127#issuecomment-2063027106

This is more about a weird -O0 -mrelax-all behavior that RISC-V now opts out, which is good on its own, even if the basis looks like a self-inflicted problem...

File: clang/test/Driver/integrated-as.c [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// XFAIL: target={{.*}}-aix{{.*}}
+// XFAIL: target={{.*}}-aix{{.*}}, target=riscv{{.*}}
```
Comment (MaskRay) [inline clang/test/Driver/integrated-as.c:1]: This causes different behaviors with the default target triple, which is probably not perfect.

I think we can change the -mrelax-all check with an explicit target triple and then add a `--target=riscv64`.

Comment (topperc) [reply clang/test/Driver/integrated-as.c:1]: What triple should i use for the -mrelax-all test?

Comment (MaskRay) [reply clang/test/Driver/integrated-as.c:1]: Perhaps the commonplace `--target=x86_64`

**Review by preames — APPROVED** [review]
LGTM

**Pull request merged**